### PR TITLE
[sharding] include round ID in dependency info

### PR DESCRIPTION
**Pull request by zjma**
### Description

Also Include the round ID (in addition to shard ID) in cross-shard edges. This is for supporting 1-blockstm-per-subblock approach, where each blockstm instance needs to know which (round,shard) to send/recv necessary txn results.

### Test Plan

Existing UTs.

### CODE REVIEW

File: execution/block-partitioner/src/sharded_block_partitioner/conflict_detector.rs [hunk 1]
```diff
@@ -51,7 +53,7 @@ impl CrossShardConflictDetector {
             } else {
                 accepted_txn_dependencies.push(self.get_deps_for_frozen_txn(
                     &txn,
-                    Arc::new(vec![]),
```
Comment (zjma) [inline execution/block-partitioner/src/sharded_block_partitioner/conflict_detector.rs:54]: This seems a bug. `get_deps_for_frozen_txn` implementation assumes the "current round rw set" vec to be of size equal to num_shards.

Comment (zjma) [reply execution/block-partitioner/src/sharded_block_partitioner/conflict_detector.rs:54]: Fixed it in this PR.

Comment (sitalkedia) [reply execution/block-partitioner/src/sharded_block_partitioner/conflict_detector.rs:54]: Why is this a bug? This needs to be empty for early return condition in `get_deps_for_frozen_txn` function.

File: execution/block-partitioner/src/sharded_block_partitioner/conflict_detector.rs [hunk 2]
```diff
@@ -72,9 +74,8 @@ impl CrossShardConflictDetector {
         current_round_rw_set_with_index: Arc<Vec<WriteSetWithTxnIndex>>,
         prev_rounds_rw_set_with_index: Arc<Vec<WriteSetWithTxnIndex>>,
     ) -> CrossShardDependencies {
-        if current_round_rw_set_with_index.is_empty() && prev_rounds_rw_set_with_index.is_empty() {
```
Comment (sitalkedia) [inline execution/block-partitioner/src/sharded_block_partitioner/conflict_detector.rs:75]: Why remove this? This makes it efficient in first round to early return.

File: execution/block-partitioner/src/sharded_block_partitioner/partitioning_shard.rs [hunk 3]
```diff
@@ -42,6 +43,7 @@ impl PartitioningShard {
             control_rx,
             result_tx,
             cross_shard_client,
+            num_rounds_finished: 0,
```
Comment (sitalkedia) [inline execution/block-partitioner/src/sharded_block_partitioner/partitioning_shard.rs:46]: better leave this partioning shard state less and not store round number here, can you pass it in the partitioning command itself?

File: types/src/block_executor/partitioner.rs [hunk 4]
```diff
@@ -9,16 +9,18 @@ pub type ShardId = usize;
 pub type TxnIndex = usize;
 
 #[derive(Debug, Clone, Eq, Hash, PartialEq, Serialize, Deserialize)]
-pub struct TxnIdxWithShardId {
+pub struct TxnIndexForSharding {
     pub txn_index: TxnIndex,
     pub shard_id: ShardId,
+    pub round: usize,
```
Comment (sitalkedia) [inline types/src/block_executor/partitioner.rs:15]: nit: May be add an alias type for round as well.

File: types/src/block_executor/partitioner.rs [hunk 5]
```diff
@@ -9,16 +9,18 @@ pub type ShardId = usize;
 pub type TxnIndex = usize;
 
 #[derive(Debug, Clone, Eq, Hash, PartialEq, Serialize, Deserialize)]
-pub struct TxnIdxWithShardId {
+pub struct TxnIndexForSharding {
```
Comment (sitalkedia) [inline types/src/block_executor/partitioner.rs:12]: How about `ShardedTxnIndex` instead?

**Review by sitalkedia — COMMENTED** [review]
Can you look into adding some tests - in this PR or as a follow up is fine.

File: execution/block-partitioner/src/sharded_block_partitioner/conflict_detector.rs (hunk 1 above)
Comment (zjma) [reply execution/block-partitioner/src/sharded_block_partitioner/conflict_detector.rs:54]: I think [here](https://github.com/aptos-labs/aptos-core/blob/703c36243b401fdba81aaa33569b54c20290307a/execution/block-partitioner/src/sharded_block_partitioner/conflict_detector.rs#L89) is the implementation that assumes "current round rw set" vec to be of size equal to num_shards.
```
            let mut current_shard_id = (self.shard_id + self.num_shards - 1) % self.num_shards; // current shard id - 1 in a wrapping fashion
            for rw_set_with_index in current_round_rw_set_with_index
                .iter()
                .take(self.shard_id)
                .rev()
                .chain(prev_rounds_rw_set_with_index.iter().rev())
            {
                if rw_set_with_index.has_write_lock(storage_location) {
                    cross_shard_dependencies.add_required_edge(
                        TxnIdxWithShardId::new(
                            rw_set_with_index.get_write_lock_txn_index(storage_location),
                            current_shard_id,
                        ),
                        storage_location.clone(),
                    );
                    break;
                }
                // perform a wrapping substraction
                current_shard_id = (current_shard_id + self.num_shards - 1) % self.num_shards;
            }
```

Imagine when `num_shards ==3, self.shard_id == 2, self.round == 2`, and we are getting deps for txns in sub-block `I`. (Other subblocks are denoted as below.)
```
round 0: A B C 
round 1: D E F
-----
round 2: G H I
```
The correct order of sub-blocks to check is `H, G, F, E, ..., B, A`.
The cursor `current_shard_id` starts with 1 which is correct.
But if `current_round_rw_set_with_index` is empty, the first sub-block being iterated in the following for-loop will be `F`. We may end up associating `F` with shard_id 1 when doing `add_required_edge`

Comment (zjma) [reply execution/block-partitioner/src/sharded_block_partitioner/conflict_detector.rs:54]: To fix it, we can either pass in empty `G, H, I` and keep current cursor logic, or fix the initial value for `current_shard_id` as well as `current_round` (introduced in this PR). I took the first approach but both are fine to me.

File: execution/block-partitioner/src/sharded_block_partitioner/conflict_detector.rs (hunk 2 above)
Comment (zjma) [reply execution/block-partitioner/src/sharded_block_partitioner/conflict_detector.rs:75]: `current_round_rw_set_with_index` won't be empty with the proposed fix (see [this](https://github.com/aptos-labs/aptos-core/pull/8826#discussion_r1241302932) thread), making it unreachable.

File: execution/block-partitioner/src/sharded_block_partitioner/partitioning_shard.rs (hunk 3 above)
Comment (zjma) [reply execution/block-partitioner/src/sharded_block_partitioner/partitioning_shard.rs:46]: sure, but it looks to me partitioning shards are already stateful? Fields `shard_id` and the rxs/txs are all tied to a partitioning session already.

File: types/src/block_executor/partitioner.rs (hunk 4 above)
Comment (zjma) [reply types/src/block_executor/partitioner.rs]: fixed

File: execution/block-partitioner/src/sharded_block_partitioner/partitioning_shard.rs (hunk 3 above)
Comment (zjma) [reply execution/block-partitioner/src/sharded_block_partitioner/partitioning_shard.rs]: fixed

File: types/src/block_executor/partitioner.rs (hunk 5 above)
Comment (zjma) [reply types/src/block_executor/partitioner.rs]: fixed

File: execution/block-partitioner/src/sharded_block_partitioner/conflict_detector.rs (hunk 1 above)
Comment (sitalkedia) [reply execution/block-partitioner/src/sharded_block_partitioner/conflict_detector.rs:54]: I see your point, can we still keep the early return condition though? It is crucial to make the first round performant by having this early return. May be we can add a special check for round == 0?

Comment (zjma) [reply execution/block-partitioner/src/sharded_block_partitioner/conflict_detector.rs:63]: fixed. added the special check outside this function to save a little bit more.

**Review by sitalkedia — APPROVED** [review]
LGTM

**Review by zekun000 — APPROVED** [review]

**Pull request merged**